keanu-project/src/main/java/io/improbable/keanu/tensor/dbl/Nd4jDoubleTensor.java

```diff
@@ -833,6 +836,75 @@ public DoubleTensor concat(int dimension, DoubleTensor... those) {
         return new Nd4jDoubleTensor(concat);
     }
 
+    /**
+     * @param dimension      the dimension to split on
+     * @param splitAtIndices the indices that the dimension to split on should be split on
+     * @return pieces of the tensor split in the order specified by splitAtIndices. To get
+     * pieces that encompasses the entire tensor, the last index in the splitAtIndices must
+     * be the length of the dimension being split on.
+     * <p>
+     * e.g A =
+     * [
+     * 1, 2, 3, 4, 5, 6
+     * 7, 8, 9, 1, 2, 3
+     * ]
+     * <p>
+     * A.split(0, [1]) gives List([1, 2, 3, 4, 5, 6])
+     * A.split(0, [1, 2] gives List([1, 2, 3, 4, 5, 6], [7, 8, 9, 1, 2, 3]
+     * <p>
+     * A.split(1, [1, 3, 6]) gives
+     * List(
+     * [1, [2, 3  , [4, 5, 6,
+     * 7]   8, 9]    1, 2, 3]
+     * )
+     */
+    @Override
+    public List<DoubleTensor> split(int dimension, int[] splitAtIndices) {
+
+        DoubleTensor duplicateTensor = this.duplicate();
+        int[] tensorShape = duplicateTensor.getShape();
+        int[] dimensionRange = TensorShape.dimensionRange(0, tensorShape.length);
+        int[] moveDimToZero = TensorShape.moveAxis(dimension, 0, dimensionRange);
+        int[] moveZeroToDim = TensorShape.moveAxis(0, dimension, dimensionRange);
+
+        DoubleTensor permutedTensor = duplicateTensor.permute(moveDimToZero).reshape(1, (int) duplicateTensor.getLength());
+
+        double[] rawBuffer = permutedTensor.asFlatDoubleArray();
+
+        List<DoubleTensor> splitTensor = new ArrayList<>();
+
+        int previousIndex = 0;
+        int rawBufferPosition = 0;
+        for (int index : splitAtIndices) {
+
+            int[] subTensorShape = Arrays.copyOf(tensorShape, tensorShape.length);
+            int subTensorLengthInDimension = index - previousIndex;
+
+            if (subTensorLengthInDimension > tensorShape[dimension] || subTensorLengthInDimension < 0) {
+                throw new IllegalArgumentException("Invalid index to split on " + index + " at " + dimension + " for tensor of shape " + Arrays.toString(tensorShape));
+            }
+
+            subTensorShape[dimension] = subTensorLengthInDimension;
+            previousIndex = index;
+
+            int subTensorLength = (int) TensorShape.getLength(subTensorShape);
+
+            double[] buffer = new double[subTensorLength];
+            System.arraycopy(rawBuffer, rawBufferPosition, buffer, 0, buffer.length);
+
+            int[] subTensorPermutedShape = TensorShape.moveAxis(dimension, 0, subTensorShape);
+
+            DoubleTensor subTensor = DoubleTensor.create(buffer, new int[]{1, subTensorLength}).reshape(subTensorPermutedShape).permute(moveZeroToDim);
```

Comment:
If a user passes 0 as an index, then `DoubleTensor.create` will throw an exception because `subTensorLength` will be 0. I think we should catch this earlier on above?

Reply:
I've updated the check

` if (subTensorLengthInDimension > tensorShape[dimension] || subTensorLengthInDimension < 0) {` to have `subTensorLengthInDimension <= 0` which should catch any split at index that causes a zero length tensor.